File: content/problems/0739-daily-temperatures/solutions/monotonic-stack/Solution.kt
class Solution {
    fun dailyTemperatures(temperatures: IntArray): IntArray {
        val answer = IntArray(temperatures.size); val stack: ArrayDeque<Int> = ArrayDeque()
        for (day in temperatures.indices) { while (stack.isNotEmpty() && temperatures[stack.last()] < temperatures[day]) { val previous = stack.removeLast(); answer[previous] = day - previous }; stack.addLast(day) }
        return answer
    }
}

File: content/problems/0739-daily-temperatures/solutions/monotonic-stack/solution.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> answer(temperatures.size()), stack;
        for (int day = 0; day < (int)temperatures.size(); ++day) { while (!stack.empty() && temperatures[stack.back()] < temperatures[day]) { int previous = stack.back(); stack.pop_back(); answer[previous] = day - previous; } stack.push_back(day); }
        return answer;
    }
};
